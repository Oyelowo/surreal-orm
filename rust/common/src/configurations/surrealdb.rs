use super::utils::get_env_vars_by_prefix;
use serde::Deserialize;
use serde_aux::prelude::deserialize_number_from_string;
use surrealdb::Datastore;

#[derive(Deserialize, Debug, Clone)]
#[serde(rename_all = "lowercase")]
pub struct SurrealdbConfigs {
    pub namespace: String,
    pub database: String,
    pub username: String,
    pub password: String,
    pub host: String,

    #[serde(deserialize_with = "deserialize_number_from_string")]
    pub port: u16,

    #[serde(default = "default_require_ssl")]
    pub require_ssl: Option<bool>,
}

fn default_require_ssl() -> Option<bool> {
    Some(false)
}

impl Default for SurrealdbConfigs {
    fn default() -> Self {
        get_env_vars_by_prefix("SURREALDB_")
    }
}

impl SurrealdbConfigs {
    pub fn get_database(self) -> anyhow::Result<Datastore> {
        // let ds = Datastore::new("tikv://127.0.0.1:8000").await?;
        todo!()
    }
}
