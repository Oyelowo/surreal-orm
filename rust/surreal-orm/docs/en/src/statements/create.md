# Create Statement

The `create` statement is used to add new entries to the SurrealDB database. It
allows you to create records with specified content and set additional
properties for the query. This documentation provides an overview of the syntax
and usage of the `create` statement.

## Table of Contents

- [Syntax](#syntax)
- [Examples](#examples)
  - [Basic Create Statement with Content Method](#basic-create-statement-with-content-method)
  - [Creating Linked Entities](#creating-linked-entities)
  - [Using the object! Macro](#using-the-object-macro)
  - [Create with Set Method](#create-with-set-method)

## Syntax

The basic syntax of the `create` statement is as follows:

```rust
create()
    .content(record_content)
    .set(set_statements)
    .return_type(return_types)
    .timeout(seconds)
    .parallel();
```

The `create` statement supports the following methods:

- `.content(record_content)`: Specifies the content of the record to be created.
- `.set(set_statements)`: Sets the values of the fields to be updated in the
  record.
- `.return_type(return_types)`: Specifies the return type for the query.
- `.timeout(seconds)`: Sets the timeout duration for the query.
- `.parallel()`: Indicates whether the query should be executed in parallel.

## Examples

### Basic Create Statement with Content Method

To create a basic record using the `create` statement, you can use the following
code:

```rust
let space_ship1 = create()
    .content(space_ship1.clone())
    .get_one(db.clone())
    .await?;
```

This code will create a new entry for `space_ship1` in the database.

### Creating Linked Entities

The `create` statement allows you to create entries that have links to other
entities. Here's an example of creating a linked entity:

```rust
let unsaved_alien = Alien {
    ...
    space_ships: LinkMany::from(vec![
        created_spaceship1.clone(),
        created_spaceship2.clone(),
        space_ship3.clone(),
    ]),
    ...
};

let created_alien_with_fetched_links = create()
    .content(unsaved_alien.clone())
    .load_link_manys()?
    .return_one(db.clone())
    .await?;
```

In this example, `unsaved_alien` is being created with links to three different
spaceships. The `.load_link_manys()` method loads the linked entities in a
single statement.

### Create Using the `object!` Macro

The `object!` macro provides a concise and type-safe way to specify values when
creating or updating records when using the `set` method. It acts as syntactic
sugar for an array of setters but provides stricter field checking compared to
using a basic struct.

Here's an example showcasing its usage:

```rust
let spaceship_id_1 = SpaceShip::create_id("spaceship1".to_string());

let space_ship1 = create::<SpaceShip>()
    .set(object!(SpaceShip {
        id: spaceship_id_1,
        name: "SpaceShip1",
        created: Utc::now(),
    }))
    .get_one(db.clone())
    .await?;
assert_eq!(space_ship1.name, "SpaceShip1");
```

Using the `object!` macro ensures all fields are provided and belong to the
specified struct. It also allows for using `parameter` or `field` as values.
This is recommended over using array of setters as shown next because of the
extra checks this provides.

Additionally, the `object_partial!` macro functions similarly but allows for
omitting some fields. This is particularly useful for update statements where
only a subset of fields need to be changed.

```rust
let updated = update::<Weapon>(id)
    .set(object_partial!(Weapon { strength: 923u64 }))
    .return_one(db.clone())
    .await?;
```

### Create with Set Method

You can use the `set` method with the `create` statement to set specific fields
of the record being created. The `set` method supports multiple approaches for
specifying the setter statements:

1. Using an array const (`&[T]`):

```rust
let space_ship2 = create::<SpaceShip>()
    .set([
        id.equal_to(spaceship_id_2),
        name.equal_to("SpaceShip2".to_string()),
        created.equal_to(Utc::now()),
    ])
    .get_one(db.clone())
    .await?;
```

2. Using a `Vec` of setter statements:

```rust
let space_ship1 = create::<SpaceShip>()
    .set(vec![
        id.equal_to(spaceship_id_1),
        name.equal_to("SpaceShip1".to_string()),
        created.equal_to(Utc::now()),
    ])
    .get_one(db.clone())
    .await?;
```

In these examples, we demonstrate different ways to use the `set` method. You
can use an array const (`[T]` or `&[T]`) or a `Vec` to provide a list of setter
statements.

This concludes the documentation for the `create` statement. Use this statement
to add new entries to the SurrealDB database with desired content and additional
properties.
