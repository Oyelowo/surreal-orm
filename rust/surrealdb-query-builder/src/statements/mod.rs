use std::fmt::{Display, Formatter};

pub(crate) mod chain;
pub(crate) mod create;
pub(crate) mod define_database;
pub(crate) mod define_event;
pub(crate) mod define_field;
pub(crate) mod define_index;
pub(crate) mod define_login;
pub(crate) mod define_namespace;
pub(crate) mod define_scope;
pub(crate) mod define_table;
pub(crate) mod define_token;
pub(crate) mod delete;
pub(crate) mod for_;
pub(crate) mod ifelse;
pub(crate) mod info;
pub(crate) mod insert;
pub(crate) mod let_;
pub(crate) mod relate;
pub(crate) mod remove_database;
pub(crate) mod remove_event;
pub(crate) mod remove_field;
pub(crate) mod remove_index;
pub(crate) mod remove_login;
pub(crate) mod remove_namespace;
pub(crate) mod remove_scope;
pub(crate) mod remove_table;
pub(crate) mod remove_token;
pub(crate) mod select;
pub(crate) mod sleep;
pub(crate) mod transaction;
pub(crate) mod update;
pub(crate) mod use_;

pub mod statements {
    pub use super::chain::{chain, QueryChain};
    pub use super::create::{create, CreateStatement};
    pub use super::define_database::{define_database, DefineDatabaseStatement};
    pub use super::define_event::{define_event, DefineEventStatement};
    pub use super::define_field::{define_field, DefineFieldStatement, FieldType};
    pub use super::define_index::{define_index, DefineIndexStatement};
    pub use super::define_login::{define_login, DefineLoginStatement};
    pub use super::define_namespace::{define_namespace, DefineNamespaceStatement};
    pub use super::define_scope::{define_scope, DefineScopeStatement};
    pub use super::define_table::{define_table, DefineTableStatement};
    pub use super::define_token::{define_token, DefineTokenStatement};
    pub use super::delete::{delete, DeleteStatement};
    pub use super::for_::{for_, For, ForCrudType};
    pub use super::ifelse::{if_, IfStatement};
    pub use super::info::{info_for, InfoStatement};
    pub use super::insert::{insert, InsertStatement, Insertables};
    pub use super::let_::{let_, LetStatement};
    pub use super::relate::{relate, RelateStatement};
    pub use super::remove_database::{remove_database, RemoveDatabaseStatement};
    pub use super::remove_event::{remove_event, RemoveEventStatement};
    pub use super::remove_field::{remove_field, RemoveFieldStatement};
    pub use super::remove_index::{remove_index, RemoveIndexStatement};
    pub use super::remove_login::{remove_login, RemoveLoginStatement};
    pub use super::remove_namespace::{remove_namespace, RemoveNamespaceStatement};
    pub use super::remove_scope::{remove_scope, RemoveScopeStatement};
    pub use super::remove_table::{remove_table, RemoveTableStatement};
    pub use super::remove_token::{remove_token, RemoveTokenStatement};
    pub use super::select::{select, SelectStatement};
    pub use super::sleep::{sleep, SleepStatement};
    pub use super::transaction::{begin_transaction, BeginTransactionStatement};
    pub use super::update::{update, UpdateStatement};
    pub use super::use_::{use_, UseStatement};
}

pub(crate) enum NamespaceOrDatabase {
    Namespace,
    Database,
}

impl Display for NamespaceOrDatabase {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let stringified = match self {
            NamespaceOrDatabase::Namespace => "NAMESPACE",
            NamespaceOrDatabase::Database => "DATABASE",
        };
        write!(f, "{}", stringified)
    }
}
