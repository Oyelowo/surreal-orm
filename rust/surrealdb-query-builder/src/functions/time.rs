/*
 * Author: Oyelowo Oyedayo
 * Email: oyelowooyedayo@gmail.com
 * Copyright (c) 2023 Oyelowo Oyedayo
 * Licensed under the MIT license
 */

// Time functions
// These functions can be used when working with and manipulating datetime values.
//
// Function	Description
// time::day()	Extracts the day as a number from a datetime
// time::floor()	Rounds a datetime down by a specific duration
// time::group()	Groups a datetime by a particular time interval
// time::hour()	Extracts the hour as a number from a datetime
// time::mins()	Extracts the minutes as a number from a datetime
// time::month()	Extracts the month as a number from a datetime
// time::nano()	Returns the number of nanoseconds since the UNIX epoch
// time::now()	Returns the current datetime
// time::round()	Rounds a datetime up by a specific duration
// time::secs()	Extracts the secs as a number from a datetime
// time::unix()	Returns the number of seconds since the UNIX epoch
// time::wday()	Extracts the week day as a number from a datetime
// time::week()	Extracts the week as a number from a datetime
// time::yday()	Extracts the yday as a number from a datetime
// time::year()	Extracts the year as a number from a datetime

use std::{fmt::Display, str::FromStr};

use crate::{
    sql::{Binding, Buildable, ToRawStatement},
    Field,
};

use super::array::Function;
use surrealdb::sql;

pub struct Datetime(sql::Value);

impl From<Datetime> for sql::Value {
    fn from(value: Datetime) -> Self {
        value.0
    }
}

impl<T: Into<sql::Datetime>> From<T> for Datetime {
    fn from(value: T) -> Self {
        let value: sql::Datetime = value.into();
        Self(value.into())
    }
}

impl From<Field> for Datetime {
    fn from(value: Field) -> Self {
        Self(value.into())
    }
}

pub struct Duration(sql::Value);

impl From<Duration> for sql::Value {
    fn from(value: Duration) -> Self {
        value.0
    }
}

impl<T: Into<sql::Duration>> From<T> for Duration {
    fn from(value: T) -> Self {
        let value: sql::Duration = value.into();
        Self(value.into())
    }
}

impl From<Field> for Duration {
    fn from(value: Field) -> Self {
        Self(value.into())
    }
}
fn day_fn(datetime: impl Into<Datetime>) -> Function {
    let binding = Binding::new(datetime.into());
    let query_string = format!("time::day({})", binding.get_param_dollarised());

    Function {
        query_string,
        bindings: vec![binding],
    }
}

#[macro_export]
macro_rules! day {
    ( $datetime:expr ) => {
        crate::functions::time::day_fn($datetime)
    };
}

pub use day;

fn floor_fn(datetime: impl Into<Datetime>, duration: impl Into<Duration>) -> Function {
    let datetime_binding = Binding::new(datetime.into());
    let duration_binding = Binding::new(duration.into());
    let query_string = format!(
        "time::floor({}, {})",
        datetime_binding.get_param_dollarised(),
        duration_binding.get_param_dollarised()
    );

    Function {
        query_string,
        bindings: vec![datetime_binding, duration_binding],
    }
}

#[macro_export]
macro_rules! floor {
    ( $datetime:expr, $duration:expr ) => {
        crate::functions::time::floor_fn($datetime, $duration)
    };
}

pub use floor;

#[derive(Debug, Clone, Copy)]
enum Interval {
    Year,
    Month,
    Hour,
    Minute,
    Second,
}

impl FromStr for Interval {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "year" => Ok(Interval::Year),
            "month" => Ok(Interval::Month),
            "hour" => Ok(Interval::Hour),
            "minute" => Ok(Interval::Minute),
            "second" => Ok(Interval::Second),
            _ => Err("Invalid interval provided. It has to be one of these: `year, month, hour, minute, second`".to_string())
        }
    }
}

impl Display for Interval {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Interval::Year => "year",
                Interval::Month => "month",
                Interval::Hour => "hour",
                Interval::Minute => "minute",
                Interval::Second => "second",
            }
        )
    }
}

enum IntervalOrField {
    Field(Field),
    Interval(Interval),
}

impl From<Field> for IntervalOrField {
    fn from(value: Field) -> Self {
        Self::Field(value)
    }
}
impl From<String> for IntervalOrField {
    fn from(value: String) -> Self {
        Self::Interval(value.parse().expect("Unable to ...todo!"))
    }
}

impl From<IntervalOrField> for sql::Value {
    fn from(value: IntervalOrField) -> Self {
        match value {
            IntervalOrField::Field(f) => f.into(),
            IntervalOrField::Interval(i) => sql::Strand::from(i.to_string()).into(),
        }
    }
}

impl From<Interval> for IntervalOrField {
    fn from(value: Interval) -> Self {
        Self::Interval(value)
    }
}

fn group_fn(datetime: impl Into<Datetime>, interval: impl Into<IntervalOrField>) -> Function {
    let datetime_binding = Binding::new(datetime.into());
    let interval_binding = Binding::new(interval.into());

    let query_string = format!(
        "time::floor({}, {})",
        datetime_binding.get_param_dollarised(),
        interval_binding.get_param_dollarised()
    );

    Function {
        query_string,
        bindings: vec![datetime_binding, interval_binding],
    }
}
// ::<crate::functions::time::IntervalOrField>
#[macro_export]
macro_rules! group {
    ( $datetime:expr, "year" ) => {
        crate::functions::time::group_fn($datetime, Interval::Year)
    };
    ( $datetime:expr, "month" ) => {
        crate::functions::time::group_fn($datetime, Interval::Month)
    };
    ( $datetime:expr, "week" ) => {
        crate::functions::time::group_fn($datetime, Interval::Week)
    };
    ( $datetime:expr, "day" ) => {
        crate::functions::time::group_fn($datetime, Interval::Day)
    };
    ( $datetime:expr, "hour" ) => {
        crate::functions::time::group_fn($datetime, Interval::Hour)
    };
    ( $datetime:expr, "minute" ) => {
        crate::functions::time::group_fn($datetime, Interval::Minute)
    };
    ( $datetime:expr, "second" ) => {
        crate::functions::time::group_fn($datetime, Interval::Second)
    };
    ( $datetime:expr, $interval:expr ) => {
        crate::functions::time::group_fn($datetime, IntervalOrField::from($interval))
    };
}

pub use group;

#[test]
fn test_day_macro_with_datetime_field() {
    let rebirth_date = Field::new("rebirth_date");
    let result = day!(rebirth_date);

    assert_eq!(result.fine_tune_params(), "time::day($_param_00000001)");
    assert_eq!(result.to_raw().to_string(), "time::day(rebirth_date)");
}
#[test]
fn test_day_macro_with_plain_datetime() {
    let dt = chrono::DateTime::<chrono::Utc>::from_utc(
        chrono::NaiveDateTime::from_timestamp(61, 0),
        chrono::Utc,
    );
    let result = day!(dt);
    assert_eq!(result.fine_tune_params(), "time::day($_param_00000001)");
    assert_eq!(
        result.to_raw().to_string(),
        "time::day('1970-01-01T00:01:01Z')"
    );
}

#[test]
fn test_floor_macro_with_datetime_field() {
    let rebirth_date = Field::new("rebirth_date");
    let duration = Field::new("duration");
    let result = floor!(rebirth_date, duration);

    assert_eq!(
        result.fine_tune_params(),
        "time::floor($_param_00000001, $_param_00000002)"
    );
    assert_eq!(
        result.to_raw().to_string(),
        "time::floor(rebirth_date, duration)"
    );
}

#[test]
fn test_floor_macro_with_plain_datetime_and_duration() {
    let dt = chrono::DateTime::<chrono::Utc>::from_utc(
        chrono::NaiveDateTime::from_timestamp(61, 0),
        chrono::Utc,
    );
    let duration = std::time::Duration::from_secs(24 * 60 * 60 * 7);
    let result = floor!(dt, duration);
    assert_eq!(
        result.fine_tune_params(),
        "time::floor($_param_00000001, $_param_00000002)"
    );
    assert_eq!(
        result.to_raw().to_string(),
        "time::floor('1970-01-01T00:01:01Z', 1w)"
    );
}

#[test]
fn test_group_macro_with_datetime_field() {
    let rebirth_date = Field::new("rebirth_date");
    let duration = Field::new("duration");
    let result = group!(rebirth_date, duration);

    assert_eq!(
        result.fine_tune_params(),
        "time::floor($_param_00000001, $_param_00000002)"
    );
    assert_eq!(
        result.to_raw().to_string(),
        "time::floor(rebirth_date, duration)"
    );
}

#[test]
fn test_group_macro_with_plain_datetime_and_duration() {
    let dt = chrono::DateTime::<chrono::Utc>::from_utc(
        chrono::NaiveDateTime::from_timestamp(61, 0),
        chrono::Utc,
    );
    let duration = std::time::Duration::from_secs(24 * 60 * 60 * 7);
    let result = group!(dt, "year");
    assert_eq!(
        result.fine_tune_params(),
        "time::floor($_param_00000001, $_param_00000002)"
    );
    assert_eq!(
        result.to_raw().to_string(),
        "time::floor('1970-01-01T00:01:01Z', 'year')"
    );
}
