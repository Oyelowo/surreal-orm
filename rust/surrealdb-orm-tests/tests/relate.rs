/*
 * Author: Oyelowo Oyedayo
 * Email: oyelowooyedayo@gmail.com
 * Copyright (c) 2023 Oyelowo Oyedayo
 * Licensed under the MIT license
 */

use insta;
use regex;
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use static_assertions::*;
use std::fmt::{Debug, Display};
use std::time::Duration;
use surrealdb::sql;
use surrealdb::{
    engine::local::{Db, Mem},
    opt::IntoResource,
    sql::Id,
    Result, Surreal,
};
use surrealdb_models::{
    book_schema, student_schema, writes_schema, Book, Student, StudentWritesBlog, StudentWritesBook,
};
use surrealdb_orm::statements::{order, relate, select};
use surrealdb_orm::*;
use test_case::test_case;
use typed_builder::TypedBuilder;

trait WhiteSpaceRemoval {
    fn remove_extra_whitespace(&self) -> String
    where
        Self: std::borrow::Borrow<str>,
    {
        let mut result = String::with_capacity(self.borrow().len());
        let mut last_char_was_whitespace = true;

        for c in self.borrow().chars() {
            if c.is_whitespace() {
                if !last_char_was_whitespace {
                    result.push(' ');
                    last_char_was_whitespace = true;
                }
            } else {
                result.push(c);
                last_char_was_whitespace = false;
            }
        }

        result
    }
}
impl WhiteSpaceRemoval for &str {}
impl WhiteSpaceRemoval for String {}

fn remove_field_from_json_string(json_string: &str, field_name: &str) -> String {
    let value: Value = serde_json::from_str(json_string).expect("Invalid JSON string");

    let updated_value = match value {
        Value::Object(mut map) => {
            map.remove(field_name);
            Value::Object(map)
        }
        Value::Array(mut vec) => {
            for element in vec.iter_mut() {
                if let Value::Object(ref mut map) = *element {
                    map.remove(field_name);
                }
            }
            Value::Array(vec)
        }
        _ => value,
    };

    serde_json::to_string(&updated_value).expect("Failed to serialize JSON value")
}

#[test]
fn can_get_structs_relations() {
    let relations_aliases = Student::get_fields_relations_aliased();
    assert_eq!(
        relations_aliases
            .into_iter()
            .map(|r| r.to_raw().build())
            .collect::<Vec<_>>()
            .join(", "),
        "->writes->book AS writtenBooks, ->writes->blog AS blogsssss"
    );
}

#[tokio::test]
async fn should_not_contain_error_when_valid_id_use_in_connection() -> SurrealdbOrmResult<()> {
    let db = Surreal::new::<Mem>(()).await.unwrap();
    db.use_ns("test").use_db("test").await.unwrap();

    let student_id1 = SurrealId::try_from("student:1").unwrap();
    let student_id2 = SurrealId::try_from("student:2").unwrap();
    let book_id1 = SurrealId::try_from("book:1").unwrap();
    let book_id2 = SurrealId::try_from("book:2").unwrap();
    let blog_id = SurrealId::try_from("blog:1").unwrap();

    let write_book = StudentWritesBook {
        time_written: Duration::from_secs(343),
        ..Default::default()
    };

    let write_book2 = StudentWritesBook {
        time_written: Duration::from_secs(923),
        ..Default::default()
    };

    let write_blog = StudentWritesBlog {
        time_written: Duration::from_secs(47),
        count: 24,
        ..Default::default()
    };

    // Student 1 writes book2
    let relation = relate(Student::with(&student_id1).writes__(Empty).book(&book_id2))
        .content(write_book.clone())
        .parallel();

    assert_eq!(relation.get_errors().len(), 0);
    let errors: Vec<String> = vec![];
    assert_eq!(relation.get_errors(), errors);

    let result = relation.load_all_links()?.return_one(db.clone()).await?;

    assert!(result
        .clone()
        .unwrap()
        .id
        .unwrap()
        .to_string()
        .starts_with("writes:"),);

    assert_eq!(
        result.clone().unwrap().in_.get_id().unwrap().to_string(),
        "student:1"
    );
    assert_eq!(
        result.clone().unwrap().out.get_id().unwrap().to_string(),
        "book:2"
    );
    assert_eq!(
        serde_json::to_string(&result).unwrap(),
        "{\"timeWritten\":{\"secs\":343,\"nanos\":0},\"count\":0}"
    );

    // Student 2 writes book1
    let relation = relate(Student::with(&student_id2).writes__(Empty).book(&book_id1))
        .content(write_book2)
        .parallel();

    assert_eq!(relation.get_errors().len(), 0);
    let errors: Vec<String> = vec![];
    assert_eq!(relation.get_errors(), errors);

    let result = relation.load_all_links()?.return_one(db.clone()).await?;

    assert!(result
        .clone()
        .unwrap()
        .id
        .unwrap()
        .to_string()
        .starts_with("writes:"),);

    assert_eq!(
        result.clone().unwrap().in_.get_id().unwrap().to_string(),
        "student:2"
    );
    assert_eq!(
        result.clone().unwrap().out.get_id().unwrap().to_string(),
        "book:1"
    );
    assert_eq!(
        // serde_jsonkresult.unwrap().time_written.to_string(),
        serde_json::to_string(&result).unwrap(),
        "{\"timeWritten\":{\"secs\":923,\"nanos\":0},\"count\":0}"
    );

    // Student 2 writes blog1
    let writes_schema::Writes { timeWritten, .. } = StudentWritesBlog::schema();
    // Using the set method
    let relation = relate(Student::with(&student_id2).writes__(Empty).blog(&blog_id))
        // .set(updater(timeWritten).equal(sql::Duration::from(Duration::from_secs(47))))
        .content(write_blog)
        .parallel();
    let writes_schema::Writes { count, .. } = StudentWritesBlog::schema();
    let relation =
        relate::<StudentWritesBlog>(Student::with(&student_id2).writes__(Empty).blog(&blog_id))
            .set(updater(count).increment_by(545))
            .set(updater(timeWritten).equal(sql::Duration::from(Duration::from_secs(47))))
            .parallel();

    assert_eq!(relation.get_errors().len(), 0);
    let errors: Vec<String> = vec![];
    assert_eq!(relation.get_errors(), errors);

    let result = relation.load_all_links()?.return_one(db.clone()).await?;

    assert!(result
        .clone()
        .unwrap()
        .id
        .unwrap()
        .to_string()
        .starts_with("writes:"));

    assert_eq!(
        result.clone().unwrap().in_.get_id().unwrap().to_string(),
        "student:2"
    );
    assert_eq!(
        result.clone().unwrap().out.get_id().unwrap().to_string(),
        "blog:1"
    );
    assert_eq!(
        // serde_jsonkresult.unwrap().time_written.to_string(),
        serde_json::to_string(&result).unwrap(),
        "{\"timeWritten\":{\"secs\":47,\"nanos\":0},\"count\":545}"
    );

    let writes_schema::Writes { timeWritten, .. } = StudentWritesBook::schema();

    let result: Vec<StudentWritesBook> = select(All)
        .from(StudentWritesBook::table_name())
        .order_by(order(timeWritten).asc())
        .return_many(db.clone())
        .await?;

    assert_eq!(result.len(), 3);
    assert_eq!(result[0].time_written, Duration::from_secs(47));
    assert_eq!(result[0].in_.get_id().unwrap().to_string(), "student:2");
    assert_eq!(result[0].out.get_id().unwrap().to_string(), "blog:1");

    assert_eq!(result[1].time_written, Duration::from_secs(343));
    assert_eq!(result[1].in_.get_id().unwrap().to_string(), "student:1");
    assert_eq!(result[1].out.get_id().unwrap().to_string(), "book:2");

    assert_eq!(result[2].time_written, Duration::from_secs(923));
    assert_eq!(result[2].in_.get_id().unwrap().to_string(), "student:2");
    assert_eq!(result[2].out.get_id().unwrap().to_string(), "book:1");

    Ok(())
}

#[test]
fn test_relation_graph_with_alias() {
    let student_id = SurrealId::try_from("student:1").unwrap();
    let book_id = SurrealId::try_from("book:2").unwrap();

    let aliased_connection = Student::with(student_id)
        .writes__(Empty)
        .book(book_id)
        .__as__(Student::aliases().writtenBooks);

    assert_eq!(
        aliased_connection.fine_tune_params(),
        "$_param_00000001->writes->$_param_00000002 AS writtenBooks"
    );

    assert_eq!(
        aliased_connection.clone().to_raw().build(),
        "student:1->writes->book:2 AS writtenBooks"
    );

    assert_eq!(aliased_connection.get_errors().len(), 0);
    let errors: Vec<String> = vec![];
    assert_eq!(aliased_connection.get_errors(), errors);
}

#[test]
fn test_recursive_edge_to_edge_connection_as_supported_in_surrealql() {
    // This allows for pattern like this:
    // -- Select all 1st, 2nd, and 3rd level people who this specific person record knows, or likes, as separate outputs
    // SELECT ->knows->(? AS f1)->knows->(? AS f2)->(knows, likes AS e3 WHERE influencer = true)->(? AS f3) FROM person:tobie;

    let student_id = SurrealId::try_from("student:1").unwrap();
    let book_id = SurrealId::try_from("book:2").unwrap();
    let likes = Table::new("likes");
    let writes = StudentWritesBook::table_name();
    let timeWritten = Field::new("timeWritten");

    // let knows = Field::writes("influencer");

    let aliased_connection = Student::with(student_id)
        .writes__(Empty)
        .writes__(Empty)
        .writes__(any_other_edges(&[writes, likes]).where_(timeWritten.less_than_or_equal(50)))
        .book(book_id)
        .__as__(Student::aliases().writtenBooks);

    assert_eq!(
        aliased_connection.fine_tune_params(),
        "$_param_00000001->writes->writes->(writes, writes, likes  WHERE timeWritten <= $_param_00000002)->$_param_00000003 AS writtenBooks"
    );

    assert_eq!(
       dbg!( aliased_connection.clone()).to_raw().build(),
        "student:1->writes->writes->(writes, writes, likes  WHERE timeWritten <= 50)->book:2 AS writtenBooks"
    );

    assert_eq!(aliased_connection.get_errors().len(), 0);
    let errors: Vec<String> = vec![];
    assert_eq!(aliased_connection.get_errors(), errors);
}

#[test]
fn test_any_edge_filter() {
    let student_id = SurrealId::try_from("student:1").unwrap();
    let book_id = SurrealId::try_from("book:2").unwrap();
    let likes = Table::new("likes");
    let wants = Table::new("wants");
    let timeWritten = Field::new("timeWritten");

    let aliased_connection = Student::with(student_id)
        .writes__(any_other_edges(&[wants, likes]).where_(timeWritten.less_than_or_equal(50)))
        .book(book_id)
        .__as__(Student::aliases().writtenBooks);

    assert_eq!(
        aliased_connection.fine_tune_params(),
        "$_param_00000001->(writes, wants, likes  WHERE timeWritten <= $_param_00000002)->$_param_00000003 AS writtenBooks"
    );

    assert_eq!(
        dbg!(aliased_connection.clone()).to_raw().build(),
        "student:1->(writes, wants, likes  WHERE timeWritten <= 50)->book:2 AS writtenBooks"
    );

    assert_eq!(aliased_connection.get_errors().len(), 0);
    let errors: Vec<String> = vec![];
    assert_eq!(aliased_connection.get_errors(), errors);
}

#[test]
fn should_contain_error_when_invalid_id_use_in_connection() {
    let student_id = SurrealId::try_from("student:1").unwrap();
    let book_id = SurrealId::try_from("book:2").unwrap();

    let write = StudentWritesBook {
        time_written: Duration::from_secs(343),
        ..Default::default()
    };

    // Book id used with student schema, while student_id used for book. This should generate
    // two errors
    let x = relate(Student::with(&book_id).writes__(Empty).book(&student_id))
        .content(write.clone())
        .return_type(ReturnType::Before)
        .parallel();

    assert_eq!(x.get_errors().len(), 2);
    assert_eq!(
        x.get_errors(),
        vec![
            "invalid id book:2. Id does not belong to table student",
            "invalid id student:1. Id does not belong to table book"
        ]
    );
}

#[tokio::test]
async fn relate_query() -> surrealdb_orm::SurrealdbOrmResult<()> {
    use surrealdb::sql::Datetime;

    let db = Surreal::new::<Mem>(()).await.unwrap();
    db.use_ns("test")
        .use_db("test")
        .await
        .expect("failed to use db");
    let student_id = SurrealId::try_from("student:1").unwrap();
    let book_id = SurrealId::try_from("book:2").unwrap();

    let write = StudentWritesBook {
        time_written: Duration::from_secs(343),
        ..Default::default()
    };

    let relate_simple = relate(Student::with(student_id).writes__(E).book(book_id)).content(write);
    assert_eq!(
        relate_simple.to_raw().build(),
        "RELATE student:1->writes->book:2 CONTENT { count: 0, timeWritten: { nanos: 0, secs: 343 } } ;"
    );

    // // You can use return one method and it just returns the single object
    // let relate_simple_object = relate_simple.return_one(db.clone()).await?;
    // // Remove id bcos it is non-deterministic i.e changes on every run
    // let relate_simple_object = remove_field_from_json_string(
    //     serde_json::to_string(&relate_simple_object)
    //         .unwrap()
    //         .as_str(),
    //     "id",
    // );
    // insta::assert_display_snapshot!(relate_simple_object);

    // // You can also use return many and it just returns the single object as an array
    // let relate_simple_array = relate_simple.return_many(db.clone()).await?;
    // let relate_simple_object = remove_field_from_json_string(
    //     serde_json::to_string(&relate_simple_object)
    //         .unwrap()
    //         .as_str(),
    //     "id",
    // );
    // insta::assert_display_snapshot!(relate_simple_object);

    Ok(())
}

#[tokio::test]
async fn relate_query_with_sub_query() -> surrealdb_orm::SurrealdbOrmResult<()> {
    let db = Surreal::new::<Mem>(()).await.unwrap();
    db.use_ns("test")
        .use_db("test")
        .await
        .expect("failed to use db");
    let student_id = SurrealId::try_from("student:1").unwrap();
    let book_id = SurrealId::try_from("book:2").unwrap();

    let write = StudentWritesBook {
        time_written: Duration::from_secs(52),
        ..Default::default()
    };
    let relate_more = relate(
        Student::with(select(All).from(Student::get_table_name()))
            .writes__(Empty)
            .book(
                select(All).from(Book::get_table_name()), // .where_(Book::schema().title.like("Oyelowo")),
            ),
    )
    .content(write)
    .return_many(db.clone())
    .await?;
    let relate_more =
        remove_field_from_json_string(serde_json::to_string(&relate_more).unwrap().as_str(), "id");

    // TODO: This returns empty array. Figure out if this is the expected behaviour
    insta::assert_display_snapshot!(relate_more);
    Ok(())
}

#[test]
fn multiplication_tests8() {
    use serde_json;

    let sur_id = SurrealId::try_from("alien:oyelowo").unwrap();
    let json = serde_json::to_string(&sur_id).unwrap();
    assert_eq!(json, "\"alien:oyelowo\"");

    let sur_id = RecordId::from(("alien", "oyelowo"));
    let json = serde_json::to_string(&sur_id).unwrap();
    assert_eq!(json, "\"alien:oyelowo\"");
}

// #[test]
// #[cfg(feature = "raw")]
// fn should_display_actual_values_in_raw_format() {
//     let student_id = SurrealId::try_from("student:1").unwrap();
//     let book_id = SurrealId::try_from("book:2").unwrap();

//     let write = StudentWritesBook {
//         time_written: Duration::from_secs(343),
//         ..Default::default()
//     };

//     let raw = relate(Student::with(&student_id).writes__(Empty).book(&book_id))
//         .content(write.clone())
//         .return_(Return::Before)
//         .parallel();

//     insta::assert_display_snapshot!(raw);
//     insta::assert_debug_snapshot!(raw.get_bindings());
// }
