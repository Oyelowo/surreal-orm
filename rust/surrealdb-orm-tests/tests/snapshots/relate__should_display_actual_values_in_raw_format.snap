---
source: surrealdb-orm-tests/tests/relate.rs
expression: raw
---
RELATE $_param_9a7d0d70->writes->$_param_f81fcf37 CONTENT $_param_154703fb  PARALLEL;
